### DISCUSSION TITLE
Is message delivery export turned on by default, and if not is there a way to do so?

### DISCUSSION THREAD

**Discussion by Qw4z1** [invertase/react-native-firebase#6499] [Q&A]
Hey!

We have started to do a lot of custom things with FCM, sending push notifications via Firebase Admin SDK. As such, we don't get access to delivery data in the Firebase Console anymore and want to use BigQuery to analyse this data.

Reading through [the documentation](https://firebase.google.com/docs/cloud-messaging/understand-delivery?platform=android#enable-message-delivery-data-export) I found that the message delivery data export needs to be enabled on the device (after asking for consent ofc).

However, I haven't seen this mentioned in the documentation for RNFirebase. Is there no support for this yet?

**Comment by mikehardy** [comment]
Hi there! I wasn't aware of that feature (firebase's API surface area is big enough, this still happens from time to time :sweat_smile: )

This looks a bit difficult because there is platform skew - but the general flow seems to be:

1 - ask users for consent, somehow, store it somehow so you may call a runtime method to enable/disable - this will be project-specific
2 - implement runtime method on react-native-firebase messaging that enables / disables it
  a - javascript method (with typescript typings) in module that call in to the native modules
  b - android native module just call directly the firebase-android-sdk method implemented in the SDK
  c - ios native module implement state variable to store enable/disable status; if enabled, background handler calls per-message method

That seems like the implementation path

Unfortunately 2c leaves out the case where the FCM has a notification block. That's tricky because it requires a notification extension helper which is not something we can just *do* for people I think. We can provide a sample / documentation but I think that's it. It may be worth crossing over into notifee.app repo and mentioning it there as well (it has lots of extension helper docs.

I think 1-2c gets you most of the way there though.

This won't be something I have time to take up but perhaps that sketch is enough to allow for a PR? I'd be happy to merge one. Our testing app is pretty good with a [README](https://github.com/invertase/react-native-firebase/blob/main/tests/README.md) showing how to drive it, and the ability to quickly write an e2e test (as in, test-driven development) that proves you are first calling a stub javascript method, then calling through to a stub method in a native module, then actually implementing things as you would expect. I estimate this would be the work of a few hours, but not days or anything ?

**Reply by Qw4z1** [reply]
Hey! Thanks for getting back to me so quickly. =) Fully aware that Firebase is a monster of an API (and it keeps growing!), so I should make it clear that I'm absolutely not expecting everything to be included here.

The steps you are describing look correct. No promises, but I'll run it by the team here at Baby Journey and see if we can get a PR up in the following weeks. We've been talking internally for a while on how we can contribute more to the open source community and this might be the perfect place for us to start. 👌

I'm going on vacation tomorrow, so there might be quite a bit of delay in my replies moving forward.

**Comment by pavan168** [comment]
Hello @mikehardy! Thanks for jotting down the implementation approach that helped us to produce the work in this PR #6529

It would be great if you can have a look sometime when you're free to see if things are in order :) 

Thanks in advance.

**Reply by mikehardy** [reply]
I see those PRs, I've queued them for review. Thanks a bunch for posting them!
